### DISCUSSION TITLE
How do I set caching headers so that the client never waits?

### DISCUSSION THREAD

**Discussion by zoul** [vercel/community#1023] [Help] [answered]
I have an API endpoint a bit like this:

```typescript
async function handler(request: NextApiRequest, response: NextApiResponse) {
  const count = await getNumberThatTakesLongToCompute();
  const oneHour = 60 * 60;
  const oneWeek = oneHour * 24 * 7;
  response.setHeader("Content-Type", "application/json");
  response.setHeader(
    "Cache-Control",
    `s-maxage=${oneHour}, stale-while-revalidate=${oneWeek}`
  );
  response.status(200).send({
    timestamp: new Date(),
    count,
  });
}
```

The response takes long to compute, so I need proper caching to avoid clients waiting. I don’t mind the clients getting a stale response. I was hoping that the cache headers I send mean that once the response is computed, clients will always get an immediate answer for a week. But that’s not the case – the response is immediate for a while, but the next day or so I have to wait again. What did I get wrong? Thanks!

**Comment by amyegan** [comment, marked as answer]
Hi @zoul! I think your `Cache-Control` settings mean that the cache will be updated after about one hour because of the current `s-maxage` value. You can learn more about `stale-while-revalidate` here: [vercel.com/docs/concepts/functions/serverless-functions/edge-caching#stale-while-revalidate](https://vercel.com/docs/concepts/functions/serverless-functions/edge-caching#stale-while-revalidate).

**Reply by zoul** [reply]
Thank you very much! I didn’t notice previously that I don’t have to specify a time value for the `stale-while-revalidate` part. And I also didn’t notice the part about custom domain invalidation:

> Every time you deploy with [a custom domain](https://vercel.com/docs/concepts/projects/custom-domains#deploying-with-your-domain), the cache for that domain is purged.

I guess that’s why the caching doesn’t last for the whole week as specified in the headers.
